### DISCUSSION TITLE
Query speed of Postgres vs REST API vs InfluxDB line protocol

### DISCUSSION THREAD

**Discussion by huzaifahj** [questdb/questdb#1136] [Ideas]
Hi there, as mentioned above is there any speed difference when querying OR inserting rows using these 3 different methods. Does the /exec method go through PostGRES anyway?

Thanks for your time.

**Comment by bluestreak01** [comment]
The fastest way to insert would be either via:

- Influxdb Line Protocol (TCP or UDP)
- PostgreSQL wire

The `/exec` is slowest for now. It doesn't support bind variables and executes `insert` via slowest path there is in the codebase. It does not use PostgreSQL wire though.

For querying, `/exec` and PostgreSQL wire are very similar. It boils down to the data format you're looking to receive. `/exec` is JSON and PostgreSQL wire is naturally via driver you use.

**Comment by huzaifahj** [comment]
That's very useful, thank you. One would assume that `insert` via /exec would be fastest as it's sort of the default method in the docs, maybe there could be a note about this.

**Comment by huzaifahj** [comment]
Can I also ask - is it normal for timeseries databases not to have a `DELETE` or `UPDATE` command as I can't find those in the docs

**Comment by bluestreak01** [comment]
I'm not sure how common the lack of `DELETE` and `UPDATE` is. We look at time series database as a tool to enable understanding of evolution of data over time. To do that data has to be immutable.

Obviously, value of data diminishes over time. To deal with that time series databases replace notion of `DELETE` with `retention`. We have partition-based retention implementation.

To deal with `UPDATE` we have `LATEST BY`. Where `update` replaces record and old state of record is gone forever, time series database should keep the history of change. `LATEST BY` enables you to retrieve the desired version of the record.
